File: compiler/feedback/type_definitions.go
package feedback

import (
	"wipple/database"
	"wipple/queries"
)

func registerTypeDefinitions() {
	register(Feedback[struct{}]{
		Id:    "duplicate-field-definition",
		Rank:  RankSyntax,
		Query: queries.DuplicateFieldDefinition,
		Render: func(render *Render, node database.Node, data struct{}) {
			render.WriteNode(node)
			render.WriteString(" is already defined in this type.")
			render.WriteBreak()
			render.WriteString("Try renaming this field.")
		},
	})

	register(Feedback[struct{}]{
		Id:    "duplicate-variant-definition",
		Rank:  RankSyntax,
		Query: queries.DuplicateVariantDefinition,
		Render: func(render *Render, node database.Node, data struct{}) {
			render.WriteNode(node)
			render.WriteString(" is already defined in this type.")
			render.WriteBreak()
			render.WriteString("Try renaming this variant.")
		},
	})
}

File: compiler/src/feedback/items/type_definitions.rs
use crate::{
    feedback::{FeedbackCtx, FeedbackRank, RegisteredFeedback},
    nodes::{DuplicateFieldDefinition, DuplicateVariantDefinition},
    queries,
};
use std::collections::BTreeSet;

pub fn register(ctx: &mut FeedbackCtx) {
    ctx.register(RegisteredFeedback::new(
        "duplicate-field-definition",
        queries::fact::<DuplicateFieldDefinition>,
        |_| FeedbackRank::Syntax,
        |(node, _)| (node.clone(), BTreeSet::new()),
        |w, (node, _)| {
            w.write_node(node);
            w.write_string(" is already defined in this type.");
            w.write_break();
            w.write_string("Try renaming this field.");
        },
    ));

    ctx.register(RegisteredFeedback::new(
        "duplicate-variant-definition",
        queries::fact::<DuplicateVariantDefinition>,
        |_| FeedbackRank::Syntax,
        |(node, _)| (node.clone(), BTreeSet::new()),
        |w, (node, _)| {
            w.write_node(node);
            w.write_string(" is already defined in this type.");
            w.write_break();
            w.write_string("Try renaming this variant.");
        },
    ));
}
